### ISSUE TITLE
plugin libs for other languages

### ISSUE THREAD

**Issue by edvorg** [google/xi-editor#574]
Hi I'm a Clojure developer and I never stop to look for a better editor than emacs.
Ideally I would prefer to configure my editor in clojure and xi architecture seems like something that could allow it in addition to good performance and safety.
The question is would it be welcome here if I submitted a PR with plugin lib for Clojure by analogy with plugin lib for Python?
From the top level I see plugin infrastructure to work in the following way:
1. clojure plugin layer - jvm process that is basically a web server, that communicates with xi using plugin protocol. clojure plugin layer is a xi plugin that includes several clojure plugins
2. clojure plugin api - allows to define plugins and submit for execution in clojure plugin layer
3. clojure networking REPL server is run by clojure plugin layer
4. clojure support on frontend - allows to communicate with existent plugins in clojure plugin layer and redefine functionality or add new clojure plugins over repl on the fly.
The solution above potentially brings full blown repl driven editing into xi where user can dynamically rewrite and redefine his configs without rerunning editor (same as it works in emacs). Another benefit is huge variety of java/clj libraries becomes available to implement xi plugins in clojure.
This is far from something complete, but if I get a green light on this issue I will dive more into designing the architecture.

**Comment by cmyr** [comment]
@edvorg I think in general plugin libraries should live outside of this repo, with the exception of the Rust one, and _possibly_ the python library, which was envisioned as (like you said) the canonical example of a third-party plugin library. It's fallen out of date lately, though, and my recent focus has shifted to just using the rust plugin library for everything.

From your architecture sketch, I'm not sure I understand #1, the extra plugin layer; since all plugin communication happens over RPC, ideally any particular lanaguage's plugin library is basically just responsible for setting up an RPC loop and providing a nice API, which will be the basis of each plugin executable. I think the best sketch of this right now is in #569, in particular the two files `rust/plugin-lib/src/global/mod.rs` and `rust/plugin-lib/src/global/dispatcher.rs`.

I would warn you, though, that the plugin API is going to be a moving target over the next few months, so there will be ongoing work to keep your library working, although I'll be happy to keep you informed of changes as they happen.

**Comment by edvorg** [comment]
Hi @cmyr 

> I'm not sure I understand your 1st point

Here is the rationale behind additional plugin layer:
1. JVM process is heavy and running all plugins in single process seems to me as better approach in this case
2. Users would expect to have have a single REPL to control all the plugins they have. Having separate plugin processes would require user to connect to each plugin separately.

Please let me know if you have a better idea in mind how to implement it given you have much better knowledge of xi.

I was thinking thinking that `1.` can be solved by using ClojureScript instead of Clojure, because node processes start faster and have smaller memory footprint. But `2.` could be a real problem.

> I would warn you, though, that the plugin API is going to be a moving target over the next few months
Sure I totally understand that editor is still under active development, I hope when it comes to implementing a plugin it would be easy to track the changes to API.

**Comment by cmyr** [comment]
Interesting.

What you're proposing doesn't _really_ fit with the current architecture; core assumes that a plugin connection is a single plugin, and I'm not sure how (for instance) we would handle routing messages to plugins hosted by an intermediary without introducing knowledge of that intermediary into xi-core.

I _could_ imagine having a 'clojure-REPL' plugin that is a collection of components, but which has its own API for managing them; to xi-core this would exist as a single plugin, but it could present commands for enabling/disabling components at runtime.

**Comment by edvorg** [comment]
While I'm just beginning to look into plugin protocol, my vision of this is to have Clojure layer declaring over rpc, what functionality does it expose (probably a list of key-bindings or functions/commands). So as functionality in Clojure layer changes due to live-reloading of the code by user, all we would have to do is to redeclare the functionality to xi-core. I understand this might be not how things work in xi right know and I'm just poking in the sky, but I'm going to look for solution.
To make a more clear example out of this.
Let's say user wants to automate writing TAGS file on file save event.
1. He has Clojure layer installed in xi. Clojure layer declares over rpc that it has a command and keybinding - execute last s-expression.
2. User connects to Clojure layer's repl.
3. User writes Clojure file with code like this

```clojure
(defplugin tags-writer
    (subscribe-to-xi-hook "save-file-hook"
         (fn [{:keys [filename]}]
              (write-tags filename))))
```

4. User hits keybinding to evaluate the s-expression above in repl. (the keybinding should be somehow exposed by clojure layer over rpc)
5. Keybinding (command) is forwarded from frontend to xi core to clojure-layer.
6. Clojure layer evaluates the form and subscribes to xi event over rpc.
7. User saves some file.
8. Xi-core notifies Clojure layer about event
9. Clojure layer executes hook that writes TAGS file.

So basically with this approach all the logic is running inside Clojure layer, while user can define such mini plugins in his configuration files and enable/update them in xi in runtime.

**Comment by cmyr** [comment]
I'm having a bit of trouble completely wrapping my head around this, but I think everything you're describing fits the current model. The only thing currently missing is key bindings; we don't support custom key bindings at the moment. We do support custom commands from plugins.

This is not a trivial project of course, and you would end up hacking on core quite a bit to add the features you need, but I'd suggest you start playing around a bit and then you'll have a better sense of what's possible. :)

**Comment by edvorg** [comment]
Yes I admit it sounds quite confusing as repl driven development is an approach that is unique to lisps. Concluding this, I think following features are required to make this thing work.
- exposing commands (supported)
- keybindings (optional, currently not supported)
- subscribe to sensitive events from xi-core (file save, file open, etc. etc.), as per example with TAGS file above this kind of functionality is essential
The last one I still have to figure out. From the quick look at `plugin.md` file at `global-plugin` branch I couldn't find any mentions about plugins being able to subscribe to events.

**Comment by edvorg** [comment]
Hi @mkeathley will make a diagram in next few days when I get a chance

**Comment by edvorg** [comment]
Haven't checked neovim as I'm not a vim user, but I'll check it out, thanks

**Comment by d4hines** [comment]
I'd like to register my interest for working on a Clojure mode. Vi-style modal editing and Clojure support are all I need from an editor.

@edvorg @mkeathley , do ya'll have any further interest in this? I assume the average Xi user values performance as first principle. In the realm of performant Clojure editing, this is a somewhat new development: https://github.com/oakmac/tree-sitter-clojure  I bet Xi + Tree Sitter could blow Atom/VS Code out of the water in terms of performance, not to mention Emacs or Cursive.

**Comment by edvorg** [comment]
Hi, sorry I wanted to work on it but couldn't allocate time due to some
higher priority things. Though I'd be willing to discuss the architecture
and help with my clojure expertise as I'm still interested in switching to
xi.
